pluginconfig/token.go

```diff
@@ -63,113 +68,189 @@ func (t TokenDataObserverConfig) WellFormed() error {
 
 // Validate checks that the observer's config is semantically correct - fields are set correctly
 // depending on the config's type
-func (t TokenDataObserverConfig) Validate() error {
+func (t *TokenDataObserverConfig) Validate() error {
+	if err := t.WellFormed(); err != nil {
+		return err
+	}
 	if t.IsUSDC() {
 		return t.USDCCCTPObserverConfig.Validate()
 	}
 	return errors.New("unknown token data observer type " + t.Type)
 }
 
-func (t TokenDataObserverConfig) IsUSDC() bool {
+func (t *TokenDataObserverConfig) IsUSDC() bool {
 	return t.Type == USDCCCTPHandlerType
 }
 
-const USDCCCTPHandlerType = "usdc-cctp"
+// MarshalJSON is a custom JSON marshaller for TokenDataObserverConfig.
+// It constructs raw map based on provided type. Custom marshaller is needed because default golang marshaller
+// doesn't marshal clashing fields of pointer embeddings even if only one pointer is present and rest are set to nil
+func (t *TokenDataObserverConfig) MarshalJSON() ([]byte, error) {
+	switch t.Type {
+	case USDCCCTPHandlerType:
+		return json.Marshal(&struct {
+			Type    string `json:"type"`
+			Version string `json:"version"`
+			*USDCCCTPObserverConfig
+		}{
+			Type:                   t.Type,
+			Version:                t.Version,
+			USDCCCTPObserverConfig: t.USDCCCTPObserverConfig,
+		})
+	default:
+		return nil, fmt.Errorf("unknown token data observer type: %q", t.Type)
+	}
+}
 
-type USDCCCTPObserverConfig struct {
-	Tokens         map[cciptypes.ChainSelector]USDCCCTPTokenConfig `json:"tokens"`
-	AttestationAPI string                                          `json:"attestationAPI"`
+// UnmarshalJSON is a custom JSON unmarshaller for TokenDataObserverConfig.
+// It first reads top-level fields, then allocates the correct embedded config pointer.
+// (only USDCCCTPObserverConfig for now) before finally unmarshalling into that pointer.
+// Custom unmarshaller is needed because default golang marshaller doesn't unmarshal clashing fields
+// (when they appear beside USDC) of pointer embeddings
+func (t *TokenDataObserverConfig) UnmarshalJSON(data []byte) error {
```

Comment:
What did exactly stop it from working? We intentionally kept everything under a single type to make deserialization simpler https://github.com/smartcontractkit/chainlink-ccip/pull/117

It was inspired by the TX types in Algorand, @winder should be able to point to the origin in their codebase

Reply:
It's an interesting point. I guess the [code here](https://github.com/algorand/go-algorand/blob/55011f93fddb181c643f8e3f3d3391b62832e7cd/data/transactions/transaction.go#L87) works because embedded types are totally distinct from one another.

If all providers require different configs and they sometimes collide then you need something like these marshal overrides. Alternatively you can prefix the colliding types, or name them rather than embedding.

Reply:
Since we've already deployed these configs, our options are limited. This might be the best way to deal with the colliding "tokens" type.